src/query/v2/plan/operator.cpp

```diff
@@ -462,6 +463,202 @@ class DistributedScanAllAndFilterCursor : public Cursor {
+      : output_symbol_(output_symbol),
+        input_cursor_(std::move(input_cursor)),
+        op_name_(op_name),
+        label_(label),
+        property_expression_pair_(property_expression_pair),
+        filter_expressions_(filter_expressions),
+        primary_key_(primary_key) {
+    ResetExecutionState();
+  }
+
+  enum class State : int8_t { INITIALIZING, COMPLETED };
+
+  using VertexAccessor = accessors::VertexAccessor;
+
+  bool MakeRequest(RequestRouterInterface &request_router, ExecutionContext &context) {
+    {
+      SCOPED_REQUEST_WAIT_PROFILE;
+      std::optional<std::string> request_label = std::nullopt;
+      if (label_.has_value()) {
+        request_label = request_router.LabelToName(*label_);
+      }
+      current_batch_ = request_router.ScanVertices(request_label);
+    }
+    current_vertex_it_ = current_batch_.begin();
+    request_state_ = State::COMPLETED;
+    return !current_batch_.empty();
+  }
+
+  bool MakeRequestSingleFrame(Frame &frame, RequestRouterInterface &request_router, ExecutionContext &context) {
+    {
+      SCOPED_REQUEST_WAIT_PROFILE;
+      std::optional<std::string> request_label = std::nullopt;
+      if (label_.has_value()) {
+        request_label = request_router.LabelToName(*label_);
+      }
+
+      // Evaluate the expressions that hold the PrimaryKey.
+      ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.request_router,
+                                    storage::v3::View::NEW);
+
+      std::vector<msgs::Value> pk;
+      MG_ASSERT(primary_key_);
+      for (auto *primary_key : *primary_key_) {
+        pk.push_back(TypedValueToValue(primary_key->Accept(evaluator)));
+      }
+
+      msgs::Label label = {.id = msgs::LabelId::FromUint(label_->AsUint())};
+
+      msgs::GetPropertiesRequest req = {.vertex_ids = {std::make_pair(label, pk)}};
+      auto get_prop_result = request_router.GetProperties(req);
+      MG_ASSERT(get_prop_result.size() <= 1);
+
+      if (get_prop_result.empty()) {
+        current_batch_ = std::vector<VertexAccessor>{};
+      } else {
+        auto properties = get_prop_result[0].props;
+        // TODO (gvolfing) figure out labels when relevant.
+        msgs::Vertex vertex = {.id = get_prop_result[0].vertex, .labels = {}};
+
+        current_batch_ = {VertexAccessor(vertex, properties, &request_router)};
+      }
+    }
+    current_vertex_it_ = current_batch_.begin();
+    request_state_ = State::COMPLETED;
+    return !current_batch_.empty();
+  }
+
+  bool Pull(Frame &frame, ExecutionContext &context) override {
+    SCOPED_PROFILE_OP(op_name_);
+
+    auto &request_router = *context.request_router;
+    while (true) {
+      if (MustAbort(context)) {
+        throw HintedAbortError();
+      }
+
+      if (request_state_ == State::INITIALIZING) {
+        if (!input_cursor_->Pull(frame, context)) {
+          return false;
+        }
+      }
```

Comment:
As we produce at most 1 row for each row from our input, I would say we always have to pull our input. That means this logic seems unnecessary. See my other comments about states and batching.